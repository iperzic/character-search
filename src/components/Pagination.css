.Pagination__arrow {
    display: inline;
    cursor: pointer;
    text-shadow: 1px 1px 1px red;
    font-size: 22px;
    padding: 8px;
}

.Pagination__arrow:hover {
    opacity: 0.7;
}

.arrow--next:after {
    content: '►';
}

.arrow--previous:after {
    content: '◄';
}

.arrow--disabled, .arrow--disabled:hover {
    opacity: 0.4;
    cursor: initial;
}
