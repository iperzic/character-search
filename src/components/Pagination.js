import React from 'react';
import PropTypes from 'prop-types';

import './Pagination.css';

const Pagination = ({ metadata, onPageChange }) => {
  const isLastPage = metadata.offset + 20 > metadata.total;
  const isFirstPage = metadata.offset - 20 < 0;
  return (
    <div className="Pagination">
      <div className={`Pagination__arrow arrow--previous ${isFirstPage ? 'arrow--disabled' : ''}`}
           onClick={() => !isFirstPage && onPageChange(metadata.offset - 20)}
           title="Previous Page"
      />
      <div className={`Pagination__arrow arrow--next ${isLastPage ? 'arrow--disabled' : ''}`}
           onClick={() => !isLastPage && onPageChange(metadata.offset + 20)}
           title="Next Page"
      />
    </div>
  );
};

Pagination.propTypes = {
  metadata: PropTypes.shape({
    count: PropTypes.number.isRequired,
    offset: PropTypes.number.isRequired,
    total: PropTypes.number.isRequired,
  }).isRequired,
  onPageChange: PropTypes.func.isRequired,
};

export default Pagination;
