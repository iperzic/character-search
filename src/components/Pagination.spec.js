import React from 'react';
import { shallow } from 'enzyme';

import Pagination from './Pagination';

describe('Pagination', () => {
  let component;
  let pageChangeSpy;
  let metadata;

  beforeEach(() => {
    pageChangeSpy = jest.fn();
    metadata = {
      count: 20,
      offset: 20,
      total: 60,
    };

    component = shallow(<Pagination metadata={metadata} onPageChange={pageChangeSpy} />)
  });

  it('should show disabled next page and disabled previous page buttons', () => {
    component.setProps({
      metadata: {
        count: 5,
        offset: 0,
        total: 5,
      }
    });

    expect(component.getElement()).toMatchSnapshot();
  });

  it('should show disabled next page and previous page buttons', () => {
    component.setProps({
      metadata: {
        count: 20,
        offset: 20,
        total: 40,
      }
    });

    expect(component.getElement()).toMatchSnapshot();
  });

  it('should show next page and disabled previous page buttons', () => {
    component.setProps({
      metadata: {
        count: 20,
        offset: 0,
        total: 40,
      }
    });

    expect(component.getElement()).toMatchSnapshot();
  });

  it('should show next page and previous page buttons', () => {
    component.setProps({
      metadata: {
        count: 20,
        offset: 20,
        total: 60,
      }
    });

    expect(component.getElement()).toMatchSnapshot();
  });

  describe('when the next page button is clicked', () => {
    describe('and the button is enabled', () => {
      it('should call the onPageChange prop', () => {
        component.find('.arrow--next').simulate('click');
        expect(pageChangeSpy).toHaveBeenCalledWith(40);
      });
    });

    describe('and the button is disabled', () => {
      it('should not call the onPageChange prop', () => {
        component.setProps({
          metadata: {
            count: 20,
            offset: 20,
            total: 40,
          }
        });
        component.find('.arrow--next').simulate('click');
        expect(pageChangeSpy).toHaveBeenCalledTimes(0);
      });
    });
  });

  describe('when the previous page button is clicked', () => {
    describe('and the button is enabled', () => {
      it('should call the onPageChange prop', () => {
        component.find('.arrow--previous').simulate('click');
        expect(pageChangeSpy).toHaveBeenCalledWith(0);
      });
    });

    describe('and the button is disabled', () => {
      it('should not call the onPageChange prop', () => {
        component.setProps({
          metadata: {
            count: 20,
            offset: 0,
            total: 40,
          }
        });
        component.find('.arrow--previous').simulate('click');
        expect(pageChangeSpy).toHaveBeenCalledTimes(0);
      });
    });
  });
});
