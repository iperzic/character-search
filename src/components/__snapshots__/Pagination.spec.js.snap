// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should show disabled next page and disabled previous page buttons 1`] = `
<div
  className="Pagination"
>
  <div
    className="Pagination__arrow arrow--previous arrow--disabled"
    onClick={[Function]}
    title="Previous Page"
  />
  <div
    className="Pagination__arrow arrow--next arrow--disabled"
    onClick={[Function]}
    title="Next Page"
  />
</div>
`;

exports[`Pagination should show disabled next page and previous page buttons 1`] = `
<div
  className="Pagination"
>
  <div
    className="Pagination__arrow arrow--previous"
    onClick={[Function]}
    title="Previous Page"
  />
  <div
    className="Pagination__arrow arrow--next arrow--disabled"
    onClick={[Function]}
    title="Next Page"
  />
</div>
`;

exports[`Pagination should show next page and disabled previous page buttons 1`] = `
<div
  className="Pagination"
>
  <div
    className="Pagination__arrow arrow--previous arrow--disabled"
    onClick={[Function]}
    title="Previous Page"
  />
  <div
    className="Pagination__arrow arrow--next"
    onClick={[Function]}
    title="Next Page"
  />
</div>
`;

exports[`Pagination should show next page and previous page buttons 1`] = `
<div
  className="Pagination"
>
  <div
    className="Pagination__arrow arrow--previous"
    onClick={[Function]}
    title="Previous Page"
  />
  <div
    className="Pagination__arrow arrow--next"
    onClick={[Function]}
    title="Next Page"
  />
</div>
`;
